source/val/Construct.h

```diff
@@ -28,29 +28,88 @@
+#include <vector>
 
 namespace libspirv {
 
+enum class ConstructType {
+  kNone,
+  /// The set of blocks dominated by a selection header, minus the set of blocks
+  /// dominated by the header's merge block
+  kSelection,
+  /// The set of blocks dominated by an OpLoopMerge's Continue Target and post
+  /// dominated by the corresponding back
+  kContinue,
+  ///  The set of blocks dominated by a loop header, minus the set of blocks
+  ///  dominated by the loop's merge block, minus the loop's corresponding
+  ///  continue construct
+  kLoop,
+  ///  The set of blocks dominated by an OpSwitch's Target or Default, minus the
+  ///  set of blocks dominated by the OpSwitch's merge block (this construct is
+  ///  only defined for those OpSwitch Target or Default that are not equal to
+  ///  the OpSwitch's corresponding merge block)
+  kCase
+};
+
 class BasicBlock;
 
 /// @brief This class tracks the CFG constructs as defined in the SPIR-V spec
 class Construct {
  public:
-  Construct(BasicBlock* header_block, BasicBlock* merge_block,
-            BasicBlock* continue_block = nullptr);
+  Construct(ConstructType type, BasicBlock* dominator,
+            BasicBlock* exit = nullptr,
+            std::vector<Construct*> constructs = {});
+
+  /// Returns the type of the construct
+  ConstructType get_type() const;
+
+  const std::vector<Construct*>& get_corresponding_constructs() const;
+  std::vector<Construct*>& get_corresponding_constructs();
+  void set_corresponding_constructs(std::vector<Construct*> constructs);
+
+  /// Returns the dominator block of the construct. This is usually the header
+  /// block or the first block of the construct.
+  const BasicBlock* get_dominator() const;
 
-  const BasicBlock* get_header() const;
-  const BasicBlock* get_merge() const;
-  const BasicBlock* get_continue() const;
+  /// Returns the dominator block of the construct. This is usually the header
+  /// block or the first block of the construct.
+  BasicBlock* get_dominator();
 
-  BasicBlock* get_header();
-  BasicBlock* get_merge();
-  BasicBlock* get_continue();
+  /// Returns the exit block of the construct. This is usually the merge block
+  /// or the last block of the construct
+  const BasicBlock* get_exit() const;
+
+  /// Returns the exit block of the construct. This is usually the merge block
+  /// or the last block of the construct
+  BasicBlock* get_exit();
+
+  /// Sets the exit block for this construct. This is useful for continue
+  /// constructs which do not know the back-edge block before during
+  /// construction
+  void set_exit(BasicBlock* exit_block);
 
  private:
-  BasicBlock* header_block_;    ///< The header block of a loop or selection
-  BasicBlock* merge_block_;     ///< The merge block of a loop or selection
-  BasicBlock* continue_block_;  ///< The continue block of a loop block
+  /// The type of the construct
+  ConstructType type_;
+
+  /// These are the constructs that are related to this construct. These
```

Comment:
Right.  I see that now in RegisterLoopMerge code.
This table (in ascii art form) is perfect.